app/src/services/tag.js

```diff
@@ -7,42 +7,196 @@ const { Tag, VersionTag } = require('../db/models');
+    }
+  },
+
+  /**
+   * @function associateTags
+   * calls createTags to create new Tag records
+   * associates new tags to the versionId
+   * @param {*} versionId The uuid id column from version table
+   * @param {object[]} tags array of tags
+   * @param {*} currentUserId uuid of current user
+   * @param {object} [etrx=undefined] An optional Objection Transaction object
+   * @returns {Promise<object>} array of all associated tags
+   * @throws The error encountered upon db transaction failure
+   */
+  associateTags: async (versionId, tags, currentUserId = SYSTEM_USER, etrx = undefined) => {
+    let trx;
+    try {
+      trx = etrx ? etrx : await Tag.startTransaction();
+      let response = [];
 
       if (tags.length) {
-        // insert/merge tag records
-        const insertTags = await Tag.query(trx)
-          .insert(tags)
-          .onConflict(['key', 'value'])
-          .merge(); // required to include id's of existing rows in result
-
-        // relate all incoming tags
-        const relateArray = insertTags.map(({ id }) => ({
-          versionId: versionId,
-          tagId: id,
-          createdBy: currentUserId
-        }));
-        response = await VersionTag.query(trx)
-          .insert(relateArray);
+        // get id's of all input tags
+        const dbTags = await service.createTags(tags, trx);
+
+        // get all currently associated tags
+        const allArray = await VersionTag.query(trx)
+          .modify('filterVersionId', versionId);
+
+        // associate remaining tags
+        const newJoins = dbTags.filter(({ id }) => {
+          return !allArray.some(({ tagId }) => tagId === id);
+        });
+        if (newJoins.length) {
+          await VersionTag.query(trx)
+            .insert(newJoins.map(({ id }) => ({
+              versionId: versionId,
+              tagId: id,
+              createdBy: currentUserId
+            })));
+        }
+        response = dbTags;
+      }
+
+      if (!etrx) await trx.commit();
+      return Promise.resolve(response);
+    } catch (err) {
+      if (!etrx && trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  /**
+   * @function dissociateTags
+   * dissociates all provided tags from a versionId
+   * @param {*} versionId The uuid id column from version table
+   * @param {string[]} keys array of tags keys (eg ['animal', 'colour'])
+   * @param {object} [etrx=undefined] An optional Objection Transaction object
+   * @returns {Promise<number>} The result of running the delete operation (number of rows deleted)
+   * @throws The error encountered upon db transaction failure
+   */
+  dissociateTags: async (versionId, keys = undefined, etrx = undefined) => {
+    let trx;
+    try {
+      trx = etrx ? etrx : await Tag.startTransaction();
+      let response = [];
+
+      // delete joining records
+      response = await VersionTag.query(trx)
+        .allowGraph('tag')
+        .withGraphJoined('tag')
+        .whereIn('key', keys)
```

Comment:
we need to ensure we have no duplicate keys, so a diff of tag records is just checking the key column